File: firebase-testing/src/com/google/gct/testing/launcher/CloudAuthenticator.java
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.gct.testing.launcher;

import com.android.annotations.Nullable;
import com.google.api.client.auth.oauth2.Credential;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.gson.GsonFactory;
import com.google.api.services.cloudresourcemanager.v3.CloudResourceManager;
import com.google.api.services.monitoring.v3.Monitoring;
import com.google.api.services.monitoring.v3.model.PointData;
import com.google.api.services.monitoring.v3.model.QueryTimeSeriesRequest;
import com.google.api.services.monitoring.v3.model.QueryTimeSeriesResponse;
import com.google.api.services.monitoring.v3.model.TimeSeriesData;
import com.google.api.services.storage.Storage;
import com.google.api.services.testing.Testing;
import com.google.api.services.testing.model.AndroidDeviceCatalog;
import com.google.api.services.toolresults.ToolResults;
import com.google.gct.login.GoogleLogin;
import com.google.gct.login.IGoogleLoginCompletedCallback;
import com.google.gct.testing.CloudTestingUtils;
import java.io.IOException;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Locale;
import java.util.TimeZone;
import kotlin.Pair;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.VisibleForTesting;

public class CloudAuthenticator {

  private static final String APPLICATION_NAME = "GCTL";

  private static CloudAuthenticator instance;

  /**
   * Global instance of the HTTP transport.
   */
  private HttpTransport myHttpTransport;
  private Credential myCredential;
  private Storage myStorage;
  private CloudResourceManager myCloudResourceManager;
  private Testing myTest;
  private Monitoring myMonitoring;
  private ToolResults myToolresults;
  private long myLastDiscoveryServiceInvocationTimestamp = -1;


  @NotNull
  public static CloudAuthenticator getInstance() {
    if (instance == null) {
      instance = new CloudAuthenticator();
    }
    return instance;
  }

  /**
   * Should be used in tests only!
   */
  @VisibleForTesting
  public static void setInstance(CloudAuthenticator testInstance) {
    instance = testInstance;
  }

  @NotNull
  public Storage getStorage() {
    prepareCredential();
    if (myStorage == null) {
      myStorage =
        new Storage.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), myCredential).setApplicationName(APPLICATION_NAME).build();
    }
    return myStorage;
  }

  public void recreateTestAndToolResults(String testBackendUrl, String toolResultsBackendUrl) {
    prepareCredential();
    myTest =
      new Testing.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), myCredential).setApplicationName(APPLICATION_NAME)
        .setRootUrl(testBackendUrl).build();
    myToolresults =
      new ToolResults.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), myCredential).setApplicationName(APPLICATION_NAME)
        .setRootUrl(toolResultsBackendUrl).build();
  }

  @NotNull
  public CloudResourceManager getCloudResourceManager() {
    prepareCredential();
    if (myCloudResourceManager == null
        || myCloudResourceManager.getRequestFactory() == null
        || myCloudResourceManager.getRequestFactory().getInitializer() != GoogleLogin.getInstance().getCredential()) {
      myCloudResourceManager =
        new CloudResourceManager.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), myCredential)
          .setApplicationName(APPLICATION_NAME).build();
    }
    return myCloudResourceManager;
  }

  /**
   * Get a test client pointing to the default (prod) backend.
   */
  @NotNull
  public Testing getTest() {
    return getTest(null);
  }

  /**
   * Get a test client pointing to the given backend.
   */
  @NotNull
  public Testing getTest(@Nullable String endpoint) {
    prepareCredential();
    if (myTest == null) {
      Testing.Builder builder =
        new Testing.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), myCredential).setApplicationName(APPLICATION_NAME);
      if (endpoint != null) {
        builder.setRootUrl(endpoint);
      }
      myTest = builder.build();
    }
    return myTest;
  }

  @NotNull
  private Monitoring getMonitoring(@Nullable String endpoint) {
    prepareCredential();
    if (myMonitoring == null) {
      Monitoring.Builder builder =
        new Monitoring.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), myCredential).setApplicationName(APPLICATION_NAME);
      if (endpoint != null) {
        builder.setRootUrl(endpoint);
      }
      myMonitoring = builder.build();
    }
    return myMonitoring;
  }

  /**
   * Get the {@link AndroidDeviceCatalog} for the given FTL {@code endpoint}.
   */
  @NotNull
  public AndroidDeviceCatalog getAndroidDeviceCatalogForEnvironment(@Nullable String endpoint, @Nullable String gcpProject)
    throws IOException {
    long currentTimestamp = System.currentTimeMillis();
    try {
      Testing.TestEnvironmentCatalog.Get getter = getTest(endpoint)
        .testEnvironmentCatalog()
        .get("ANDROID");
      getter.setProjectId(gcpProject);
      getter.getRequestHeaders().set("X-Goog-User-Project", gcpProject);
      AndroidDeviceCatalog catalog = getter
        .execute()
        .getAndroidDeviceCatalog();
      if (catalog.getVersions().isEmpty() || catalog.getModels().isEmpty() || catalog.getRuntimeConfiguration().getLocales().isEmpty()
          || catalog.getRuntimeConfiguration().getOrientations().isEmpty()) {
        showDeviceCatalogError("Android device catalog is empty for some dimensions", currentTimestamp);
      }
      return catalog;
    }
    finally {
      myLastDiscoveryServiceInvocationTimestamp = currentTimestamp;
    }
  }

  /**
   * Get the {@link AndroidDeviceCatalog} for the default (prod) FTL backend.
   */
  @Nullable
  public AndroidDeviceCatalog getAndroidDeviceCatalog() {
    try {
      return getAndroidDeviceCatalogForEnvironment(null, null);
    }
    catch (IOException e) {
      showDeviceCatalogError("Exception while getting Android device catalog\n\n" + e.getMessage(), System.currentTimeMillis());
      return null;
    }
  }

  @NotNull
  private QueryTimeSeriesResponse queryMonitoring(@NotNull String endpoint, @NotNull String project, @NotNull String queryString)
    throws IOException {
    Monitoring monitoring = getMonitoring(endpoint);
    QueryTimeSeriesRequest request = new QueryTimeSeriesRequest()
      .setQuery(queryString)
      .setPageSize(200);
    return monitoring.projects().timeSeries()
      .query(project, request)
      .execute();
  }

  /**
   * Returns a pair of usage and limit numbers of quota in minutes for the endPoint and project, null if not available.
   *
   * @param endpoint end point of the monitoring backend, effective only for the first calling
   * @param project  name of the cloud project
   */
  @Nullable
  public Pair<Long, Long> getQuotaUsageAndLimit(@NotNull String endpoint, @NotNull String project) {
    Calendar calendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
    calendar.setTimeInMillis(CloudTestingUtils.getTimestampAtMidnightInPT(Instant.now()));
    // Sets up the beginning date of the query interval.
    String date = String.format(Locale.US, "d'%d/%d/%d 7:00'",
                                calendar.get(Calendar.YEAR),
                                calendar.get(Calendar.MONTH) + 1,
                                calendar.get(Calendar.DAY_OF_MONTH));

    try {
      QueryTimeSeriesResponse usageResponse = queryMonitoring(
        endpoint,
        project,
        "fetch consumer_quota | metric 'serviceruntime.googleapis.com/quota/rate/net_usage'\n" +
        "| filter metric.quota_metric==\"testing.googleapis.com/direct_access/blaze_physical_minutes\" " +
        "|| metric.quota_metric==\"testing.googleapis.com/direct_access/spark_physical_minutes\"\n" +
        "| within " + date
      );
      // Response does not has enough data to determine usage.
      if (usageResponse.size() < 2) {
        return null;
      }
      long usageNumber = sumNumbers(usageResponse);
      QueryTimeSeriesResponse limitResponse = queryMonitoring(
        endpoint,
        project,
        "fetch consumer_quota\n" +
        "| metric 'serviceruntime.googleapis.com/quota/limit'\n" +
        "| filter metric.limit_name==\"BlazePhysicalDeviceDirectAccessMinutesPerDayPerProject\"" +
        "|| metric.limit_name==\"SparkPhysicalDeviceDirectAccessMinutesPerDayPerProject\"\n" +
        "| within " + date
      );
      long limitNumber = findNumber(limitResponse);
      // Response does not has enough data to determine usage limit.
      if (usageResponse.size() < 2) {
        return null;
      }
      return new Pair<>(usageNumber, limitNumber);
    }
    catch (Exception e) {
      // TODO: Surface errors in the UI.
      CloudTestingUtils.showErrorMessage(null, "Error retrieving remaining quotas",
                                         "Failed to retrieve remaining quotas! Please try again later.\n" + e.getLocalizedMessage());
      return null;
    }
  }

  private long findNumber(@NotNull QueryTimeSeriesResponse item) {
    TimeSeriesData timeSeriesData = (TimeSeriesData)((ArrayList<?>)item.get("timeSeriesData")).get(0);
    PointData pointData = timeSeriesData.getPointData().get(0);
    return pointData.getValues().get(0).getInt64Value();
  }

  private long sumNumbers(@NotNull QueryTimeSeriesResponse item) {
    TimeSeriesData timeSeriesData = (TimeSeriesData)((ArrayList<?>)item.get("timeSeriesData")).get(0);
    return timeSeriesData.getPointData().stream().mapToLong((a) -> a.getValues().get(0).getInt64Value()).sum();
  }

  private void showDeviceCatalogError(String errorMessageSuffix, long currentTimestamp) {
    // The error should be reported just once per burst of invocations.
    if (currentTimestamp - myLastDiscoveryServiceInvocationTimestamp > 1000L) { // If more than a second has passed.
      CloudTestingUtils.showErrorMessage(null, "Error retrieving android device catalog",
                                         "Failed to retrieve available firebase devices! Please try again later.\n" + errorMessageSuffix);
    }
  }

  @NotNull
  public ToolResults getToolresults() {
    prepareCredential();
    if (myToolresults == null) {
      myToolresults =
        new ToolResults.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), myCredential).setApplicationName(APPLICATION_NAME)
          .build();
    }
    return myToolresults;
  }

  public void prepareCredential() {
    if (myHttpTransport == null) {
      myHttpTransport = createHttpTransport();
    }
    GoogleLogin googleLogin = GoogleLogin.getInstance();
    if (myCredential == null || myCredential != googleLogin.getCredential()) {
      if (!authorize()) {
        throw new RuntimeException("Failed to authorize to Google Cloud! Please check if you set the correct user account.");
      }
      myCredential = googleLogin.getCredential();
    }
  }

  @NotNull
  private HttpTransport createHttpTransport() {
    try {
      return new NetHttpTransport();
    }
    catch (Exception e) {
      System.err.println(e.getMessage());
      throw new RuntimeException("Failed to acquire HTTP transport for Google Cloud Storage!");
    }
  }

  /**
   * Authorizes the installed application to access user's protected data.
   */
  public static boolean authorize() {
    final GoogleLogin googleLogin = GoogleLogin.getInstance();
    Credential credential = googleLogin.getCredential();
    if (credential == null) {
      googleLogin.logIn(null, (IGoogleLoginCompletedCallback)null);
      credential = googleLogin.getCredential();
      return credential != null;
    }
    return true;
  }

  public static boolean isUserLoggedIn() {
    return GoogleLogin.getInstance().getCredential() != null;
  }
}

File: firebase-testing/src/com/google/gct/testing/launcher/CloudAuthenticator.kt
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.gct.testing.launcher

import com.google.api.client.http.HttpTransport
import com.google.api.client.http.javanet.NetHttpTransport
import com.google.api.client.json.gson.GsonFactory
import com.google.api.services.cloudbilling.Cloudbilling
import com.google.api.services.cloudresourcemanager.v3.CloudResourceManager
import com.google.api.services.monitoring.v3.Monitoring
import com.google.api.services.monitoring.v3.model.PointData
import com.google.api.services.monitoring.v3.model.QueryTimeSeriesRequest
import com.google.api.services.monitoring.v3.model.QueryTimeSeriesResponse
import com.google.api.services.monitoring.v3.model.TimeSeriesData
import com.google.api.services.storage.Storage
import com.google.api.services.testing.Testing
import com.google.api.services.testing.model.AndroidDeviceCatalog
import com.google.api.services.toolresults.ToolResults
import com.google.gct.login.LoginState
import com.google.gct.login2.GoogleLoginService
import com.google.gct.login2.LoginFeature
import com.google.gct.testing.CloudTestingUtils
import com.google.services.firebase.FirebaseLoginFeature
import com.intellij.openapi.components.Service
import com.intellij.openapi.components.service
import java.io.IOException
import java.time.Instant
import java.util.Calendar
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.CompletableFuture
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

private const val APPLICATION_NAME = "GCTL"

@Service
class CloudAuthenticator(scope: CoroutineScope) {
  /** Global instance of the HTTP transport. */
  private var myHttpTransport: HttpTransport? = null
    get() = field ?: NetHttpTransport().also { field = it }

  private var myStorage: Storage? = null
  private var myCloudbilling: Cloudbilling? = null
  private var myCloudResourceManager: CloudResourceManager? = null
  private var myTest: Testing? = null
  private var myMonitoring: Monitoring? = null
  private var myToolresults: ToolResults? = null
  private var myLastDiscoveryServiceInvocationTimestamp: Long = -1

  init {
    scope.launch {
      fun reset() {
        myHttpTransport = null
        myStorage = null
        myCloudResourceManager = null
        myTest = null
        myMonitoring = null
        myToolresults = null
      }

      if (service<GoogleLoginService>().useOldVersion) {
        service<LoginState>().loginStatus.collect { reset() }
      } else {
        service<GoogleLoginService>().activeUserFlow.collect { reset() }
      }
    }
  }

  private val firebaseFeature = LoginFeature.feature<FirebaseLoginFeature>()

  val storage: Storage
    get() {
      return myStorage
        ?: Storage.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), firebaseFeature.credential())
          .setApplicationName(APPLICATION_NAME)
          .build()
          .also { myStorage = it }
    }

  fun recreateTestAndToolResults(testBackendUrl: String?, toolResultsBackendUrl: String?) {
    myTest =
      Testing.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), firebaseFeature.credential())
        .setApplicationName(APPLICATION_NAME)
        .setRootUrl(testBackendUrl)
        .build()
    myToolresults =
      ToolResults.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), firebaseFeature.credential())
        .setApplicationName(APPLICATION_NAME)
        .setRootUrl(toolResultsBackendUrl)
        .build()
  }

  val cloudbilling: Cloudbilling
    get() {
      return myCloudbilling
        ?: Cloudbilling.Builder(
          myHttpTransport,
          GsonFactory.getDefaultInstance(),
          firebaseFeature.credential(),
        ).setApplicationName(APPLICATION_NAME).build().also { myCloudbilling = it }
    }

  val cloudResourceManager: CloudResourceManager
    get() {
      return myCloudResourceManager
        ?: CloudResourceManager.Builder(
            myHttpTransport,
            GsonFactory.getDefaultInstance(),
            firebaseFeature.credential(),
          )
          .setApplicationName(APPLICATION_NAME)
          .build()
          .also { myCloudResourceManager = it }
    }

  /** Get a test client pointing to the default (prod) backend. */
  val test: Testing
    get() = getTest(null)

  /** Get a test client pointing to the given backend. */
  private fun getTest(endpoint: String?): Testing {
    return myTest
      ?: Testing.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), firebaseFeature.credential())
        .setApplicationName(APPLICATION_NAME)
        .apply {
          if (endpoint != null) {
            setRootUrl(endpoint)
          }
        }
        .build()
        .also { myTest = it }
  }

  private fun getMonitoring(endpoint: String?): Monitoring {
    return myMonitoring
      ?: Monitoring.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), firebaseFeature.credential())
        .setApplicationName(APPLICATION_NAME)
        .apply {
          if (endpoint != null) {
            setRootUrl(endpoint)
          }
        }
        .build()
        .also { myMonitoring = it }
  }

  /** Get the [AndroidDeviceCatalog] for the given FTL `endpoint`. */
  @Throws(IOException::class)
  fun getAndroidDeviceCatalogForEnvironment(
    endpoint: String?,
    gcpProject: String?,
  ): AndroidDeviceCatalog {
    val currentTimestamp = System.currentTimeMillis()
    try {
      val getter = getTest(endpoint).testEnvironmentCatalog()["ANDROID"]
      getter.setProjectId(gcpProject)
      getter.requestHeaders["X-Goog-User-Project"] = gcpProject
      val catalog = getter.execute().androidDeviceCatalog
      if (
        catalog.versions.isEmpty() ||
          catalog.models.isEmpty() ||
          catalog.runtimeConfiguration.locales.isEmpty() ||
          catalog.runtimeConfiguration.orientations.isEmpty()
      ) {
        showDeviceCatalogError(
          "Android device catalog is empty for some dimensions",
          currentTimestamp,
        )
      }
      return catalog
    } finally {
      myLastDiscoveryServiceInvocationTimestamp = currentTimestamp
    }
  }

  /** Get the [AndroidDeviceCatalog] for the default (prod) FTL backend. */
  val androidDeviceCatalog: AndroidDeviceCatalog?
    get() {
      try {
        return getAndroidDeviceCatalogForEnvironment(null, null)
      } catch (e: IOException) {
        showDeviceCatalogError(
          """
  Exception while getting Android device catalog

  ${e.message}
  """
            .trimIndent(),
          System.currentTimeMillis(),
        )
        return null
      }
    }

  fun isBillingEnabled(cloudProject: String):Boolean =
      cloudbilling.projects().getBillingInfo("projects/$cloudProject").execute().billingEnabled

  @Throws(IOException::class)
  private fun queryMonitoring(
    endpoint: String,
    project: String,
    queryString: String,
  ): QueryTimeSeriesResponse {
    val monitoring = getMonitoring(endpoint)
    val request = QueryTimeSeriesRequest().setQuery(queryString).setPageSize(200)
    return monitoring.projects().timeSeries().query(project, request).execute()
  }

  /**
   * Returns remaining quota in minutes for the endPoint and project, -1 if not available.
   *
   * @param endpoint end point of the monitoring backend, effective only for the first calling
   * @param project name of the cloud project
   */
  fun getQuotaUsageAndLimit(endpoint: String, project: String): Pair<Long, Long>? {
    val calendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
    calendar.timeInMillis = CloudTestingUtils.getTimestampAtMidnightInPT(Instant.now())
    // Sets up the beginning date of the query interval.
    val date =
      String.format(
        Locale.US,
        "d'%d/%d/%d 7:00'",
        calendar[Calendar.YEAR],
        calendar[Calendar.MONTH] + 1,
        calendar[Calendar.DAY_OF_MONTH],
      )

    try {
      val usageResponse =
        queryMonitoring(
          endpoint,
          project,
          """fetch consumer_quota | metric 'serviceruntime.googleapis.com/quota/rate/net_usage'
| filter metric.quota_metric=="testing.googleapis.com/direct_access/blaze_physical_minutes" || metric.quota_metric=="testing.googleapis.com/direct_access/spark_physical_minutes"
| within $date""",
        )
      // Response does not has enough data to determine usage.
      if (usageResponse.size < 2) {
        return null
      }
      val usageNumber = sumNumbers(usageResponse)
      val limitResponse =
        queryMonitoring(
          endpoint,
          project,
          """
              fetch consumer_quota
              | metric 'serviceruntime.googleapis.com/quota/limit'
              | filter metric.limit_name=="BlazePhysicalDeviceDirectAccessMinutesPerDayPerProject"|| metric.limit_name=="SparkPhysicalDeviceDirectAccessMinutesPerDayPerProject"
              | within $date
              """
            .trimIndent(),
        )
      val limitNumber = findNumber(limitResponse)
      // Response does not has enough data to determine usage limit.
      if (usageResponse.size < 2) {
        return null
      }
      return Pair(usageNumber, limitNumber)
    } catch (e: Exception) {
      // TODO: Surface errors in the UI.
      return null
    }
  }

  private fun findNumber(item: QueryTimeSeriesResponse): Long {
    val timeSeriesData = (item["timeSeriesData"] as ArrayList<*>?)!![0] as TimeSeriesData
    val pointData = timeSeriesData.pointData[0]
    return pointData.getValues()[0].int64Value
  }

  private fun sumNumbers(item: QueryTimeSeriesResponse): Long {
    return (item["timeSeriesData"] as ArrayList<*>)
      .stream()
      .mapToLong { timeSeriesData ->
        (timeSeriesData as TimeSeriesData)
          .pointData
          .stream()
          .mapToLong { a: PointData -> a.getValues()[0].int64Value }
          .sum()
      }
      .sum()
  }

  private fun showDeviceCatalogError(errorMessageSuffix: String, currentTimestamp: Long) {
    // The error should be reported just once per burst of invocations.
    if (
      currentTimestamp - myLastDiscoveryServiceInvocationTimestamp > 1000L
    ) { // If more than a second has passed.
      CloudTestingUtils.showErrorMessage(
        null,
        "Error retrieving android device catalog",
        "Failed to retrieve available firebase devices! Please try again later.\n$errorMessageSuffix",
      )
    }
  }

  val toolresults: ToolResults
    get() =
      myToolresults
        ?: ToolResults.Builder(myHttpTransport, GsonFactory.getDefaultInstance(), firebaseFeature.credential())
          .setApplicationName(APPLICATION_NAME)
          .build()
          .also { myToolresults = it }

  fun prepareCredential() {
    if (!firebaseFeature.isLoggedIn()) {
      if (!authorize()) {
        throw RuntimeException(
          "Failed to authorize to Google Cloud! Please check if you set the correct user account."
        )
      }
    }
  }

  /** Authorizes the installed application to access user's protected data. */
  fun authorize(): Boolean {
    if (!firebaseFeature.isLoggedIn()) {
      val complete = CompletableFuture<Nothing>()
      firebaseFeature.logInAsync { complete.complete(null) }
      complete.get()
    }
    return firebaseFeature.isLoggedIn()
  }

  companion object {

    @JvmStatic
    val instance: CloudAuthenticator
      get() = service<CloudAuthenticator>()

    @JvmStatic
    @Deprecated("Just check the status directly")
    val isUserLoggedIn: Boolean
      get() = LoginFeature.feature<FirebaseLoginFeature>().isLoggedIn()
  }
}
